File: reference/array-hash/303-Range-Sum-Query-Immutable.go
type NumArray struct {
	prefixNum []int
}

func Constructor(nums []int) NumArray {
	prefixSum := make([]int, len(nums))
	prefixSum[0] = nums[0]

	for i := 1; i < len(nums); i++ {
		prefixSum[i] = prefixSum[i-1] + nums[i]
	}

	return NumArray{
		prefixNum: append([]int{}, prefixSum...),
	}
}

func (this *NumArray) SumRange(left int, right int) int {
	if left == 0 {
		return this.prefixNum[right]
	}
	return this.prefixNum[right] - this.prefixNum[left-1]
}

File: reference/array-hash/303-Range-Sum-Query-Immutable.py
class NumArray:

    def __init__(self, nums: List[int]):
        self.prefixNum = [0] * len(nums)
        self.prefixNum[0] = nums[0]

        for i in range(1, len(nums)):
            self.prefixNum[i] = self.prefixNum[i - 1] + nums[i]

    def sumRange(self, left: int, right: int) -> int:
        if left == 0:
            return self.prefixNum[right]
        return self.prefixNum[right] - self.prefixNum[left - 1]